### PR TITLE
Packaging Refactor

### LINKED ISSUE #3227: labextension uninstall does not work on incompatible extensions
I am trying to work on an extension, and after upgrading jupyterlab I've discovered that another extension [@jupyterlab/github](https://github.com/jupyterlab/jupyterlab-github)  is now out of date and its dependencies do not resolve properly. 

I discovered that disabling this extension still doesn't allow jupyterlab to run. 

So, I attempted to uninstall it with 

```bash
jupyter labextension uninstall @jupyterlab/github
```

And that raised an error finishing with: 

```
JupyterLab              Extension            Package
>=0.12.0-0 <0.13.0-0    >=0.11.0-0 <0.12.0-0 @jupyterlab/application
>=0.12.4-0 <0.13.0-0    >=0.11.0-0 <0.12.0-0 @jupyterlab/apputils
>=0.12.0-0 <0.13.0-0    >=0.11.0-0 <0.12.0-0 @jupyterlab/coreutils
>=0.12.0-0 <0.13.0-0    >=0.11.0-0 <0.12.0-0 @jupyterlab/docmanager
>=0.12.0-0 <0.13.0-0    >=0.11.0-0 <0.12.0-0 @jupyterlab/filebrowser
>=0.51.0-0 <0.52.0-0    >=0.50.0-0 <0.51.0-0 @jupyterlab/services
```

…the incompatibility of which is exactly why I wanted to uninstall the extension.

Is there any way around this? @ian-r-rose happens to be awesome and has already released a version with upgraded dependencies. But it should be possible to get back to a state where you can have a working version of jupyterlab even if one of your extensions is out-of-date. I'm sure there's a way to do that, but I can't easily figure it out…

### LINKED ISSUE #3231: Unusual error when installing labextension
I was trying to install the xkcd labextension and ran:
```
jupyter labextension install jupyterlab_xkcd
```
Now, I later found that the name I was giving was wrong - it isnt `jupyterlab_xkcd` but it's `@jupyterlab/xkcd-extension`

But the error I got was weird (it gives a "Not a gzipped file" error):
```python
> node node-version-check.js
> npm pack jupyterlab_xkcd
npm ERR! Darwin 16.7.0
npm ERR! argv ".nvm/versions/node/v7.4.0/bin/node" ".nvm/versions/node/v7.4.0/bin/npm" "pack" "jupyterlab_xkcd"
npm ERR! node v7.4.0
npm ERR! npm  v4.0.5
npm ERR! code E404

npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/jupyterlab_xkcd
npm ERR! 404 
npm ERR! 404  'jupyterlab_xkcd' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! Please include the following file with any support request:
npm ERR!     miniconda3/share/jupyter/lab/extensions/temp/npm-debug.log
Traceback (most recent call last):
  File "miniconda3/lib/python3.5/tarfile.py", line 1629, in gzopen
    t = cls.taropen(name, mode, fileobj, **kwargs)
  File "miniconda3/lib/python3.5/tarfile.py", line 1605, in taropen
    return cls(name, mode, fileobj, **kwargs)
  File "miniconda3/lib/python3.5/tarfile.py", line 1470, in __init__
    self.firstmember = self.next()
  File "miniconda3/lib/python3.5/tarfile.py", line 2279, in next
    tarinfo = self.tarinfo.fromtarfile(self)
  File "miniconda3/lib/python3.5/tarfile.py", line 1082, in fromtarfile
    buf = tarfile.fileobj.read(BLOCKSIZE)
  File "miniconda3/lib/python3.5/gzip.py", line 274, in read
    return self._buffer.read(size)
  File "miniconda3/lib/python3.5/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
  File "miniconda3/lib/python3.5/gzip.py", line 461, in read
    if not self._read_gzip_header():
  File "miniconda3/lib/python3.5/gzip.py", line 409, in _read_gzip_header
    raise OSError('Not a gzipped file (%r)' % magic)
OSError: Not a gzipped file (b'0 ')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "miniconda3/bin/jupyter-labextension", line 11, in <module>
    sys.exit(main())
  File "miniconda3/lib/python3.5/site-packages/jupyter_core/application.py", line 266, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "miniconda3/lib/python3.5/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/labextensions.py", line 174, in start
    super(LabExtensionApp, self).start()
  File "miniconda3/lib/python3.5/site-packages/jupyter_core/application.py", line 255, in start
    self.subapp.start()
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/labextensions.py", line 64, in start
    for arg in self.extra_args]
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/labextensions.py", line 64, in <listcomp>
    for arg in self.extra_args]
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/commands.py", line 137, in install_extension
    return IOLoop.instance().run_sync(func)
  File "miniconda3/lib/python3.5/site-packages/tornado/ioloop.py", line 458, in run_sync
    return future_cell[0].result()
  File "miniconda3/lib/python3.5/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "miniconda3/lib/python3.5/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/commands.py", line 191, in install_extension_async
    data = _read_package(pjoin(target, fname))
  File "miniconda3/lib/python3.5/site-packages/jupyterlab/commands.py", line 1073, in _read_package
    tar = tarfile.open(target, "r:gz")
  File "miniconda3/lib/python3.5/tarfile.py", line 1575, in open
    return func(name, filemode, fileobj, **kwargs)
  File "miniconda3/lib/python3.5/tarfile.py", line 1633, in gzopen
    raise ReadError("not a gzip file")
tarfile.ReadError: not a gzip file
```

### LINKED ISSUE #3264: Better extension validation on install
We should use our buildutils to verify an extension when installing, and make sure it has the extension paths:
cf https://travis-ci.org/jupyterlab/mimerender-cookiecutter/builds/304858014#L5350

### LINKED ISSUE #3279: Extension link failure
I am getting an error on master when running `jupyter labextension link .`:
```bash
> /home/ian/anaconda3/envs/jupyter/bin/jlpm pack .
yarn pack v1.3.2
error Couldn't find a package.json file in "/tmp/tmp96ed_isl"
info Visit https://yarnpkg.com/en/docs/cli/pack for documentation about this command.
Traceback (most recent call last):
  File "/home/ian/anaconda3/envs/jupyter/bin/jupyter-labextension", line 11, in <module>
    load_entry_point('jupyterlab', 'console_scripts', 'jupyter-labextension')()
  File "/home/ian/anaconda3/envs/jupyter/lib/python3.6/site-packages/jupyter_core/application.py", line 266, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/home/ian/anaconda3/envs/jupyter/lib/python3.6/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/home/ian/jupyter/jupyterlab/jupyterlab/labextensions.py", line 179, in start
    super(LabExtensionApp, self).start()
  File "/home/ian/anaconda3/envs/jupyter/lib/python3.6/site-packages/jupyter_core/application.py", line 255, in start
    self.subapp.start()
  File "/home/ian/jupyter/jupyterlab/jupyterlab/labextensions.py", line 94, in start
    for arg in self.extra_args]
  File "/home/ian/jupyter/jupyterlab/jupyterlab/labextensions.py", line 94, in <listcomp>
    for arg in self.extra_args]
  File "/home/ian/jupyter/jupyterlab/jupyterlab/commands.py", line 204, in link_package
    handler.link_package(path)
  File "/home/ian/jupyter/jupyterlab/jupyterlab/commands.py", line 469, in link_package
    info = self._extract_package(path, tempdir)
  File "/home/ian/jupyter/jupyterlab/jupyterlab/commands.py", line 953, in _extract_package
    raise ValueError('Extension failed to pack')
ValueError: Extension failed to pack
```

### PR DESCRIPTION
**Pull request by blink1073**
- Separates core and dev modes.  They do not share file paths and are now semantically different.  The `jupyterlab/` directory is no longer part of the `yarn` workspaces, so we don't need our previous `egg_info` workaround due to the symlinks.  This also makes it much clearer what the assets are and who they belong to (e.g. there is no more `package.app.json`).
- Uses an updated `setupbase.py` from [jupyter-packaging](https://github.com/jupyter/jupyter-packaging) to ensure build assets.  We always ensure a core mode, and ensure that the core mode matches the current version before publishing. 
- If installing in `develop` mode, we run the equivalent of `jlpm install`.  We also automatically ensure that the application is built before running `jupyter lab --dev-mode`.
- Adds caching of the miniconda and python packages, but starting with a fresh conda environment each time
-  Improves lab extension CLI and fixes some bugs as well:  Fixes #3264.  Fixes #3231.  Fixes #3227.  Fixes #3279.

### CODE REVIEW

**Comment by blink1073** [comment]
Back to wip because we appear to have an issue with wheel building.

**Comment by blink1073** [comment]
The issue was that we were looking for the `data_files` before they had been built.  We now specify them declaratively so we can fetch them after the node build step.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -82,14 +82,13 @@ jupyter serverextension enable --py jupyterlab
 
 Notes:
 
-* The `jlpm` command is a JupyterLab-provided, locked version of the [yarn](https://yarnpkg.com/en/).  If you have `yarn` installed already, you can use
-the `yarn` command when developing, and it will use the local version of `yarn`
- in `jupyterlab/yarn.js` when run in the repository or a built application
+* The `jlpm` command is a JupyterLab-provided, locked version of the [jlpm](https://yarnpkg.com/en/).  If you have `jlpm` installed already, you can use
```
Comment (ian-r-rose) [inline CONTRIBUTING.md]: This looks like an overly-aggressive find/replace for `yarn`-->`jlpm`

File: buildutils/src/create-package.ts [hunk 2]
```diff
@@ -37,5 +37,5 @@ inquirer.prompt(questions).then(answers => {
   data.name = name;
   data.description = description;
   utils.writePackageData(jsonPath, data);
-  utils.run('jlpm run integrity');
+  utils.run('npm run integrity');
```
Comment (ian-r-rose) [inline buildutils/src/create-package.ts]: Can you comment on the places where `npm` must still be used? I have found that confusing (it looks like it is most of the scripts in buildutils?).

File: buildutils/src/update-core-mode.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+utils.writePackageData(path.join(staging, 'package.json'), data);
+
+// Update our index file and webpack file.
+fs.copySync('./dev_mode/index.js', './jupyterlab/staging/index.js');
+fs.copySync('./dev_mode/webpack.config.js',
+            './jupyterlab/staging/webpack.config.js');
+
+
+// Update the jlpm.lock file.
+utils.run('jlpm', { cwd: staging });
```
Comment (ian-r-rose) [inline buildutils/src/update-core-mode.ts]: But not this one.

File: docs/extensions_user.md [hunk 4]
```diff
@@ -4,9 +4,9 @@ JupyterLab extensions add functionality to the JupyterLab application.
 They can provide new file viewer types, launcher activities, and new Notebook
 output renderers for example.
 
-### Installing Node.js and npm
+### Installing Node.j
```
Comment (ian-r-rose) [inline docs/extensions_user.md]: Typo: `Node.j` --> `Node.js`

File: setup.py [hunk 5]
```diff
@@ -139,21 +137,20 @@
 }
 
 
-if 'setuptools' in sys.modules:
-    setup_args.update(setuptools_args)
+# Because of this we do not need a MANIFEST.in
```
Comment (vidartf) [inline setup.py]: Then the file can (and should?) be removed?

Comment (blink1073) [reply setup.py]: Yes, it is removed in this PR.

Comment (vidartf) [reply setup.py]: Oh, I must have missed it in the changed files overview.

**Pull request merged**